plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/edit.tsx

```diff
@@ -15,88 +18,120 @@ import { ProductGalleryThumbnailsBlockSettings } from './block-settings';
+				setOverflowState( overflow );
+			} );
 
-		return () => {
-			resizeObserver.disconnect();
-		};
-	}, [ thumbnailSize ] ); // Re-run when thumbnailSize changes as it affects layout
+			// Observe both the scrollable element and its parent for size changes
+			resizeObserver.observe( scrollableElement );
+			if ( scrollableElement.parentElement ) {
+				resizeObserver.observe( scrollableElement.parentElement );
+			}
 
-	const thumbnailSizeValue = Number( thumbnailSize.replace( '%', '' ) );
+			// Initial check
+			const overflow = checkOverflow( scrollableElement );
+			setOverflowState( overflow );
+
+			return () => {
+				resizeObserver.disconnect();
+			};
+		}, [ thumbnailSize ] ); // Re-run when thumbnailSize changes as it affects layout
 
-	const className = clsx(
-		'wc-block-product-gallery-thumbnails',
-		`wc-block-product-gallery-thumbnails--thumbnails-size-${ thumbnailSizeValue }`,
-		{
-			'wc-block-product-gallery-thumbnails--overflow-right':
-				overflowState.right,
-			'wc-block-product-gallery-thumbnails--overflow-bottom':
-				overflowState.bottom,
-		}
-	);
-	const blockProps = useBlockProps( { className } );
+		const thumbnailSizeValue = Number( thumbnailSize.replace( '%', '' ) );
+		const className = clsx(
+			'wc-block-product-gallery-thumbnails',
+			`wc-block-product-gallery-thumbnails--thumbnails-size-${ thumbnailSizeValue }`,
+			{
+				'wc-block-product-gallery-thumbnails--overflow-right':
+					overflowState.right,
+				'wc-block-product-gallery-thumbnails--overflow-bottom':
+					overflowState.bottom,
+			}
+		);
+		const blockProps = useBlockProps( { className } );
 
-	return (
-		<div { ...blockProps }>
-			<InspectorControls>
-				<PanelBody>
-					<ProductGalleryThumbnailsBlockSettings
-						attributes={ attributes }
-						setAttributes={ setAttributes }
-					/>
-				</PanelBody>
-			</InspectorControls>
-			<div
-				ref={ scrollableRef }
-				className="wc-block-product-gallery-thumbnails__scrollable"
-			>
-				{ [ ...Array( 10 ).keys() ].map( ( index ) => {
-					return (
+		return (
+			<>
+				<InspectorControls>
+					<PanelBody>
+						<ProductGalleryThumbnailsBlockSettings
+							attributes={ attributes }
+							setAttributes={ setAttributes }
+						/>
+					</PanelBody>
+				</InspectorControls>
+				{ renderThumbnails && (
+					<div { ...blockProps }>
 						<div
-							className="wc-block-product-gallery-thumbnails__thumbnail"
-							key={ index }
+							ref={ scrollableRef }
+							className="wc-block-product-gallery-thumbnails__scrollable"
 						>
-							<img
-								className="wc-block-product-gallery-thumbnails__thumbnail__image"
-								src={ `${ WC_BLOCKS_IMAGE_URL }block-placeholders/product-image-gallery.svg` }
-								alt=""
-							/>
+							{ productThumbnails.map(
+								( { src, alt }, index ) => {
+									return (
+										<div
+											className="wc-block-product-gallery-thumbnails__thumbnail"
+											key={ index }
```

Comment:
In the interest of performance, can we not generate a key value in our utility functions instead of relying on index, especially in cases where we have the image data?

Reply:
If we use real images it doesn't really matter because images don't change under any condition. While if we use placeholders all have the same `src` and `alt` and cannot be distinguished. So I guess using `index` in both cases is justified here.